xarray/core/dataset.py

```diff
@@ -1110,16 +1122,90 @@ def relevant_keys(mapping):
                 # dim already exists
                 raise ValueError('Existing dimensions are not valid choices '
                                  'for the dim argument in sel_points')
+            # replace the array dim
+            dim_coord = dim.copy()
+            dim = dim.name
 
         if not utils.is_scalar(dim) and not isinstance(dim, DataArray):
-            dim = as_variable(dim, name='points')
+            name = 'points' if not hasattr(dim, 'name') else dim.name
+            dim_coord = as_variable(dim, name=name)
+            dim = name
+
+        indexers_dict = dict(indexers)
+        non_indexed_dims = set(self.dims) - indexer_dims
+
+        sel_coords = OrderedDict()
+        if dim_coord is not None:
+            sel_coords[dim] = dim_coord
+
+        # Add sliced versions of coordinates that themselves depend on other dimensions
+        # which may also have been indexed
+        for c in self.coords:
+            if c in indexers_dict:
+                selection = take(self.coords[c], indexers_dict[c])
+                sel_coords[c] = (dim, selection)
+
+            elif any(d in indexer_dims for d in self.coords[c].dims):
+                # If the coord variable depends on an indexed dimension, slice it
+                var = self.coords[c]
+                coord_dim = var.dims[0]  # should just be one?
+                selection = take(var, indexers_dict[coord_dim])
+
+                sel_coords[c] = (dim, selection)
+            else:
+                sel_coords[c] = self.coords[c]
+
+        variables = OrderedDict()
+
+        for name in data_vars:
+            var = self.variables[name]
+            # Transpose the var to ensure that the indexed dims come first
+            # These dims will be collapsed in the output.
+            # To avoid edge cases in numpy want to transpose to ensure the indexed dimensions are first
+            # However transpose is not lazy, so want to avoid using it for dask case (??)
+            var = var.transpose(*(list(d for d in indexer_dims if d in var.dims) +
```

Comment:
You might transpose the entire dataset once, instead of transposing each variable in turn. It should do basically the same thing.